REGRESSION(264617@main): fast/forms/ios/file-upload-panel-accept.html is a constant timeout

hortont424 opened the pull request:
#### 6caf4c76fb2e97c1be972addb3270dfc4026c18e
<pre>
REGRESSION(264617@main): fast/forms/ios/file-upload-panel-accept.html is a constant timeout
<a href="https://bugs.webkit.org/show_bug.cgi?id=258901">https://bugs.webkit.org/show_bug.cgi?id=258901</a>
rdar://110879270

Reviewed by Simon Fraser.

* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxy::commitLayerTree):
The explicit commit introduced in 264617@main caused multiple bits of unexpected
fallout, breaking a layout test and bypassing commit-coalescing optimizations
in UIKit.

Instead of explicitly committing, just hold on to the IOSurface ports until
the commit arrives in the render server. (We can&apos;t hold on to the whole
transaction because the ports are consumed when applying the backing store
to its layer).

Canonical link: <a href="https://commits.webkit.org/265825@main">https://commits.webkit.org/265825@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7072d3b6b828dee61318c09822e9b3c2cb50444c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/11995 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/12343 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/12644 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/13741 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/11578 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/12014 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/14753 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/12357 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/14276 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/12159 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/12969 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/10148 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/14153 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/10261 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/10884 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/18013 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/11338 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/11043 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/14215 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/11530 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/9490 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/10744 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/15069 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/1336 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/31/builds/11385 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas reviewed (commented):
Looks good to me!

aprotyas commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
```diff
@@ -152,10 +152,25 @@
+
+    // Keep IOSurface send rights alive until the commit makes it to the render server, otherwise we will
+    // prematurely drop the only reference to them, and `inUse` will be wrong for a brief window.
+    [CATransaction addCommitHandler:[sendRights = WTFMove(sendRights)]() { } forPhase:kCATransactionPhasePostSynchronize];
```
btw, in [a recent PR](https://github.com/WebKit/WebKit/pull/15552) I had to `mutable` a handler because it copied the capture args rather than moving (when converting from a C++ lambda to an ObjC block). That may or may not be needed here?...

whsieh replied:
Hm...I don't _think_ it's necessary here; in your other PR, the `makeBlockPtr` call (which wraps `Block_copy` in non-ARC) was turning the move into a copy, but in this case it's directly converting the lambda to a block.

(Might be good to double check, however).

nt1m reviewed (commented):
Can we unskip the test: https://searchfox.org/wubkat/rev/12cb9f518180ff4a469eff2481efdf907e4cf63c/LayoutTests/platform/ios/TestExpectations#3724 ?

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
```diff
@@ -152,10 +152,25 @@
+    Vector<MachSendRight> sendRights;
+    for (auto& transaction : transactions) {
+        // commitLayerTreeTransaction consumes the incoming buffers, so we need to grab them first.
+        for (auto& [layerID, properties] : transaction.first.changedLayerProperties()) {
```
Can we use an existing traversal to gather them, rather than adding a new one?

hortont424 replied:
I don't think it's worth smearing this touches-a-single-function patch across 4 classes for that, but maybe? I'm going to run with this one for now to fix the regression and we can consider making it messier later :)

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
```diff
@@ -152,10 +152,25 @@
+    for (auto& transaction : transactions) {
+        // commitLayerTreeTransaction consumes the incoming buffers, so we need to grab them first.
+        for (auto& [layerID, properties] : transaction.first.changedLayerProperties()) {
+            const RemoteLayerBackingStoreProperties *backingStoreProperties = properties->backingStoreProperties.get();
```
auto

hortont424 committed:
REGRESSION(264617@main): fast/forms/ios/file-upload-panel-accept.html is a constant timeout
https://bugs.webkit.org/show_bug.cgi?id=258901
rdar://110879270

Reviewed by Simon Fraser.

* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxy::commitLayerTree):
The explicit commit introduced in 264617@main caused multiple bits of unexpected
fallout, breaking a layout test and bypassing commit-coalescing optimizations
in UIKit.

Instead of explicitly committing, just hold on to the IOSurface ports until
the commit arrives in the render server. (We can't hold on to the whole
transaction because the ports are consumed when applying the backing store
to its layer).

Canonical link: https://commits.webkit.org/265825@main
